dev/tools/gen_keycodes/lib/key_data.dart

```diff
@@ -0,0 +1,367 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:path/path.dart' as path;
+import 'package:meta/meta.dart';
+
+import 'package:gen_keycodes/utils.dart';
+
+/// The data structure used to manage keyboard key entries.
+///
+/// The main constructor parses the given input data into the data structure.
+///
+/// The data structure can be also loaded and saved to JSON, with the
+/// [KeyData.fromJson] constructor and [toJson] method, respectively.
+class KeyData {
+  /// Parses the input data given in from the various data source files,
+  /// populating the data structure.
+  ///
+  /// None of the parameters may be null.
+  KeyData(
+    String chromiumHidCodes,
+    String androidKeyboardLayout,
+    String androidKeyCodeHeader,
+    String androidNameMap,
+  )   : assert(chromiumHidCodes != null),
+        assert(androidKeyboardLayout != null),
+        assert(androidKeyCodeHeader != null),
+        assert(androidNameMap != null) {
+    _nameToAndroidScanCodes = _readAndroidScanCodes(androidKeyboardLayout);
+    _nameToAndroidKeyCode = _readAndroidKeyCodes(androidKeyCodeHeader);
+    final Map<String, List<dynamic>> dynamicNames = json.decode(androidNameMap).cast<String, List<dynamic>>();
+    _nameToAndroidName = dynamicNames.map<String, List<String>>((String key, List<dynamic> value) {
+      return MapEntry<String, List<String>>(key, value.cast<String>());
+    });
+    data = _readHidEntries(chromiumHidCodes);
+  }
+
+  /// Parses the given JSON data and populates the data structure from it.
+  KeyData.fromJson(String jsonFile) {
+    final Map<String, dynamic> contentMap = json.decode(jsonFile);
+    data = <Key>[];
+    for (String key in contentMap.keys) {
+      data.add(Key.fromJson(key, contentMap[key]));
+    }
+  }
+
+  /// Converts the data structure into a JSON file that can be parsed by
+  /// [KeyData.fromJson].
+  String toJson() {
```

Comment:
`toJson()` should return a json encodable object like a map, not a json encoded string.

Reply:
same here, though in general toJson is a special method

Reply:
OK, made it more normal and moved the encoding to the main function.